### PR TITLE
Add header propagation logic to client

### LINKED ISSUE #729: Context propagation support
Make use of server `Header` field to propagate context across workflows.

### PR DESCRIPTION
**Pull request by shreyassrivatsan**
* Add ContextPropagator interface that allows propagating selected context information across a workflow
* Add context propagator to code on the StartWorkflow, Activity paths. 

Fixes #729

### CODE REVIEW

File: internal/context.go [hunk 1]
```diff
@@ -419,3 +426,15 @@ func (c *valueCtx) Value(key interface{}) interface{} {
 	}
 	return c.Context.Value(key)
 }
+
+func spanFromContext(ctx Context) opentracing.Span {
+	val := ctx.Value(activeSpanKey)
+	if sp, ok := val.(opentracing.Span); ok {
+		return sp
+	}
+	return nil
```
Comment (mfateev) [inline internal/context.go]: Shouldn't we return NOOP Span instead?

File: internal/worker.go [hunk 2]
```diff
@@ -163,6 +164,10 @@ type (
 		// Optional: worker graceful shutdown timeout
 		// default: 0s
 		WorkerStopTimeout time.Duration
+
+		// Optional: Sets ContextPropagators that allows users to control the context information passed through a workflow
+		// default: no ContextPropagators
+		ContextPropagators []ContextPropagator
```
Comment (mfateev) [inline internal/worker.go]: may be make the jaeger or open tracing implementation (if exists) a default?

File: internal/internal_worker_base.go [hunk 3]
```diff
@@ -81,11 +82,12 @@ type (
 		IsReplaying() bool
 		MutableSideEffect(id string, f func() interface{}, equals func(a, b interface{}) bool) encoded.Value
 		GetDataConverter() encoded.DataConverter
+		GetContextPropagators() []ContextPropagator
 	}
 
 	// WorkflowDefinition wraps the code that can execute a workflow.
 	workflowDefinition interface {
-		Execute(env workflowEnvironment, input []byte)
+		Execute(env workflowEnvironment, header *shared.Header, input []byte)
```
Comment (mfateev) [inline internal/internal_worker_base.go]: I would add header to the workflowEnvironment

File: internal/tracer_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	"testing"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/stretchr/testify/assert"
+	jaeger_config "github.com/uber/jaeger-client-go/config"
+	"go.uber.org/cadence/.gen/go/shared"
+)
+
+func TestTracingContextPropagator(t *testing.T) {
+	config := jaeger_config.Configuration{}
```
Comment (mfateev) [inline internal/tracer_test.go]: So open tracing doesn't have its own implementation?

File: internal/workflow.go [hunk 5]
```diff
@@ -846,6 +847,13 @@ func WithDataConverter(ctx Context, dc encoded.DataConverter) Context {
 	return ctx1
 }
 
+// WithContextPropagators adds ContextPropagators to the context.
+func WithContextPropagators(ctx Context, contextPropagators []ContextPropagator) Context {
```
Comment (mfateev) [inline internal/workflow.go]: I don't think it should be public

File: internal/tracer_test.go (hunk 4 above)
Comment (shreyassrivatsan) [reply internal/tracer_test.go]: Unfortunately not. Had to pull on a tracing implementation just for the purpose of a test. They only have a noop tracer that does not implement the Inject, Extract methods we are interested in.

File: internal/worker.go (hunk 2 above)
Comment (shreyassrivatsan) [reply internal/worker.go]: Will do that in the next change when I actually add the tracing support. Enabling tracing will add the default tracing context propagator.

File: internal/context.go (hunk 1 above)
Comment (shreyassrivatsan) [reply internal/context.go]: This is a copy of the open tracing implementation. Returns a nil in absence of an opentracing Span.

**Comment by shreyassrivatsan** [comment]
@mfateev added a bunch more tests and responded to comments..

File: internal/context.go (hunk 1 above)
Comment (mfateev) [reply internal/context.go]: Does it mean that all Span using code always checks for this nil before using a Span?

Comment (shreyassrivatsan) [reply internal/context.go]: Yes, but generally code (even RPC) doesnt need to deal with spans directly. They do something like.. 
```
span = opentracing.SpanFromContext(ctx)
newSpan = tracer.CreateSpan(ChildOf(span))
defer newSpan.Finish()
```

So the tracer is expected to handle that span can actually be nil..

File: internal/internal_workflow_testsuite.go [hunk 6]
```diff
@@ -404,11 +406,16 @@ func (env *testWorkflowEnvironmentImpl) executeWorkflowInternal(delayStart time.
 		panic(err)
 	}
 	env.workflowDef = workflowDefinition
+	header := &shared.Header{
+		Fields: map[string][]byte{
+			"test-header": []byte("test-data"),
+		},
+	}
```
Comment (mfateev) [inline internal/internal_workflow_testsuite.go]: What is the purpose of this header for workflow testing?

File: internal/internal_workflow_testsuite.go [hunk 7]
```diff
@@ -463,6 +470,11 @@ func (env *testWorkflowEnvironmentImpl) executeActivity(
 		},
 		ActivityType: *activityType,
 		Input:        input,
+		Header: &shared.Header{
+			Fields: map[string][]byte{
+				testHeader: []byte("test-data"),
+			},
+		},
```
Comment (mfateev) [inline internal/internal_workflow_testsuite.go]: What is the purpose of this header for activity testing?

File: internal/internal_workflow_testsuite.go [hunk 8]
```diff
@@ -52,6 +52,7 @@ const (
 	defaultTestWorkflowTypeName = "default-test-workflow-type-name"
 	defaultTestDomainName       = "default-test-domain-name"
 	workflowTypeNotSpecified    = "workflow-type-not-specified"
+	testHeader                  = "test-header"
```
Comment (mfateev) [inline internal/internal_workflow_testsuite.go]: I think customers should be able to inject any headers into the workflow under the test.

File: workflow/context_propagator.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	// ContextPropagator is an interface that determines what information from
+	// context to pass along
+	ContextPropagator = internal.ContextPropagator
+)
+
+// ContextKey to be used for the StringMapPropagator
+type ContextKey string
+
+// stringMapPropagator propagates the list of keys across a workflow,
+// interpreting the payloads as strings. This is provided as a sample for other
```
Comment (mfateev) [inline workflow/context_propagator.go]: Why not interpreting payloads as bytes?

File: internal/workflow_testsuite.go [hunk 10]
```diff
@@ -68,6 +68,11 @@ type (
 		runFn        func(args mock.Arguments)
 		waitDuration func() time.Duration
 	}
+
+	// TestContextPropagator propagates test keys across a workflow
```
Comment (mfateev) [inline internal/workflow_testsuite.go]: Why do we need this? Can we just use the public one?

File: internal/internal_workflow_testsuite.go (hunk 6 above)
Comment (shreyassrivatsan) [reply internal/internal_workflow_testsuite.go]: Removed, made this configurable so user can use on tests..

File: internal/internal_workflow_testsuite.go (hunk 7 above)
Comment (shreyassrivatsan) [reply internal/internal_workflow_testsuite.go]: Removed, made this configurable so user can use on tests..

File: internal/internal_workflow_testsuite.go (hunk 8 above)
Comment (shreyassrivatsan) [reply internal/internal_workflow_testsuite.go]: Removed, made this configurable so user can use on tests..

File: internal/workflow_testsuite.go (hunk 10 above)
Comment (shreyassrivatsan) [reply internal/workflow_testsuite.go]: Done

File: workflow/context_propagator.go (hunk 9 above)
Comment (shreyassrivatsan) [reply workflow/context_propagator.go]: It's just a sample. The tracing propagator already has a sample of marshalling payloads into bytes, so wanted this sample to deal with strings..

Comment (mfateev) [reply workflow/context_propagator.go]: I'm not sure that sample should be part of the core library. May be move it into a samples repo?

Comment (shreyassrivatsan) [reply workflow/context_propagator.go]: Would you prefer if I leave it as internal and use it just for test purposes?

Comment (shreyassrivatsan) [reply workflow/context_propagator.go]: I'll move this to be a local struct that is used only by my test..

**Pull request merged**